### PR TITLE
New Block: Description List

### LINKED ISSUE #4880: Block for description list
## Issue Overview
[Description lists](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dl) (`<dl>`)  are useful to implement a glossary or to display metadata as a list of key-value pairs. I think it would be interesting to add them as block to offer editors the right markup for their glossaries, product/services specifications, etc. directly in the post content.

This tag seems underused, but I see a lot of use cases. Why not to making it more visible, easy to use and in the same time to promote the use of the right markup? 

## Possible Solution
As suggested by @aduth in the [Slack channel](https://wordpress.slack.com/archives/C02QB2JS7/p1517860272000084), this block could be seen as a variation of the "List" block.

## Poll
Do you use this tag frequently? (if not: why)

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

Fixes #4880

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dl

This is a block that I think would have been more difficult to implement without the use of light blocks. With light blocks, we're able to use proper semantic tags in the editor, which also allow it to be styled more easily.

To do: it might be could to handle enter/delete from rich text.

Screenshot 2020-03-10 at 12 24 37

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by Soean** [comment]
Fixes #4880

**Review by mcsf — COMMENTED** [review]
Definitely cool to see how neatly we can build this now.

Now, can we make this more ergonomic? Suggestions:
- `dd`: <kbd>Enter</kbd> twice to start a new `dt` (not too different from
- `dt`: <kbd>Space</kbd> (like List blocks) to convert to `dd`
- `dd`: (not sure about this one) <kbd>Backspace</kbd> to to convert to `dt`
- Add transforms between those two types
- Because DLs are visually so clean, I wonder if we need visual cues to let the user know that they are starting a new `dt` or `dd` — but what?

**Comment by ellatrix** [comment]
Yes, I left that as a to do :D

> To do: it might be could to handle enter/delete from rich text.

I think we should also take into account that you can have multiple `dt` elements (terms) and multiple `dd` elements (details). In other words, you could have multiple entries explaining a term, and you can have multiple terms for one explanation. See https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dl#Multiple_terms_single_description.

Maybe we could have `Enter` once create a block of the same type, and `Enter` twice create a block of the other type. This will require more context of surrounding blocks though.

**Comment by mcsf** [comment]
> I think we should also take into account that you can have multiple `dt` elements (terms) and multiple `dd` elements (details).

Good reminder. Additionally, this to me reinforces the need for us to _somehow_ address: "Because DLs are visually so clean, I wonder if we need visual cues to let the user know that they are starting a new dt or dd"

> Maybe we could have Enter once create a block of the same type, and Enter twice create a block of the other type. This will require more context of surrounding blocks though.

Might work, yeah.

File: packages/block-library/src/description-list/index.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+export const name = 'core/description-list';
+
+export const settings = {
+	title: __( 'Description List' ),
+	description: __( 'List groups of terms and descriptions.' ),
+	keywords: [ __( 'list' ), __( 'definitions' ), __( 'terms' ) ],
+	category: 'layout',
+	supports: {
+		className: false,
```
Comment (ZebulanStanphill) [inline packages/block-library/src/description-list/index.js:20]: Why is this set to `false`?

For consistency, I think we should add color controls to this block if we [add them to the List block](https://github.com/WordPress/gutenberg/pull/21387).

File: packages/block-library/src/description-list/index.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ */
+import edit from './edit';
+import save from './save';
+
+export const name = 'core/description-list';
+
+export const settings = {
+	title: __( 'Description List' ),
+	description: __( 'List groups of terms and descriptions.' ),
+	keywords: [ __( 'list' ), __( 'definitions' ), __( 'terms' ) ],
```
Comment (ZebulanStanphill) [inline packages/block-library/src/description-list/index.js:17]: I don't think `list` is a necessary keyword when the title already includes that word. Maybe `glossary` would be a better keyword to use here?

File: packages/block-library/src/description-list/index.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * Internal dependencies
+ */
+import edit from './edit';
+import save from './save';
+
+export const name = 'core/description-list';
+
+export const settings = {
+	title: __( 'Description List' ),
+	description: __( 'List groups of terms and descriptions.' ),
```
Comment (ZebulanStanphill) [inline packages/block-library/src/description-list/index.js:16]:
```suggestion
	description: __( 'List of terms and their descriptions.' ),
```

File: packages/block-library/src/description-details/index.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	parent: [ 'core/description-details' ],
+	description: __( 'The details of a term.' ),
+	category: 'layout',
+	supports: {
+		className: false,
+		lightBlockWrapper: true,
+	},
+	edit,
+	save,
+	attributes: {
```
Comment (ZebulanStanphill) [inline packages/block-library/src/description-details/index.js:25]: This, the name, and the category should be moved to a `block.json`. (Same with other blocks being added in this PR.)

**Comment by paaljoachim** [comment]
Quite some time has passed since there was any movement here.
What can do to help move this PR forward?

**Comment by peroyomas** [comment]
Currently both the [WHATWG](https://html.spec.whatwg.org/multipage/grouping-content.html#the-dl-element) and the [W3C](https://www.w3.org/TR/html52/grouping-content.html#the-dl-element) HTML5 standards support grouping `dt` and `dd` tags using the `div` element ([link to the issue were it was first discussed](https://github.com/whatwg/html/issues/1937)), which is useful on several cases where a single concept uses more than one `dt` or `dd` tag and stuff like CSS and semantics becomes tricky. I think that would be nice to automatically wrap groups of _x_ number of` dt` followed by _x_ number of `dd` in `div` tags, even if in a list the `dt` and `dd` elements aren't directly followed by `dt` or `dd` slibings respectively.

**Comment by ellatrix** [comment]
Wrapping in divs is tricky, because then we somehow need to ensure that inside that div dt elements come first and the dd elements follow. If these are mixed, then there's no point to group them.

**Comment by ghost** [comment]
@ellatrix Are you still working on this? And could you summarize anything that's left to be done before this can be merged?

**Comment by hans2103** [comment]
Quite some time has passed since there was any movement here.
What can do to help move this PR forward?

Link to the first website ever
https://info.cern.ch/hypertext/WWW/TheProject.html
See source code... even a description list over there!

**Review by hans2103 — APPROVED** [review]
tested successfully


@youknowriad
@talldan
@Soean
@ajitbohra
@jorgefilipecosta 

you are assigned as reviewer. Please take action on it and test this PR. We're missing the Description List

**Comment by hacknug** [comment]
@hans2103 just use a plugin. It is obvious they don't see any value on adding this block. Issues were opened during the alpha or beta stages of Gutenberg around 2018 iirc and people have been waiting ever since. I don't even work with WordPress anymore after seeing all of this lol

**Comment by ahaubold** [comment]
Hi everyone,

I’ve been looking for a way to use a definition list in Gutenberg and came across this ticket – it looks like a promising and well-considered proposal.

Since <dl> support is quite valuable for structured content (e.g., glossaries, term lists, accessibility use cases), I was wondering if there are any plans to pick this up again or move it forward.

I understand that priorities shift over time, but I’d love to see this become part of Gutenberg if possible. Is there anything that might help to revive the discussion around this block?

Thanks again to everyone involved for your effort!

**Comment by chubes4** [comment]
Let's land this!

**Comment by SteveRyan-ASU** [comment]
For anyone following this earlier implementation, a current Description List block proposal is now open in #81728.

The newer PR builds on the ideas explored here while updating the implementation for current Gutenberg APIs. It uses separate Description List, Description Term, and Description Detail blocks; preserves flexible multiple-term and multiple-description sequences; and includes toolbar and Tab/Shift+Tab transforms between term and detail blocks.

It is currently up to date with `trunk`, and all automated checks are passing. Review and feedback would be very welcome on #81728.
